File: apiCaller.js
export async function fetchGeographicPoints() {
    try {
        const response = await fetch('https://elementz.rguhack.uk/pointsOfInterest', {
            method: 'GET',
            mode: 'no-cors', // This will make it a "opaque" response
        });
        if (!response.ok) throw new Error(`HTTP error! Status: ${response.status}`);

        const points = await response.json(); // Expecting an array of objects like [{ lat, lon, name, description }]
        
        console.log(`Fetched ${points.length} geographic points of interest:`);
        points.forEach((point, index) => {
            console.log(`Point ${index + 1}:`, point);
        });

        return points;
    } catch (error) {
        console.error('Error fetching geographic points:', error);
        return [];
    }
}

export async function addMarkersFromAPI(viewer) {
    const points = await fetchGeographicPoints();

    // Create a ScreenSpaceEventHandler to handle hover events
    const handler = new Cesium.ScreenSpaceEventHandler(viewer.scene.canvas);
    handler.setInputAction((movement) => {
        const pickedObject = viewer.scene.pick(movement.endPosition);
        if (Cesium.defined(pickedObject)) {
            // Show the info on hover
            const { name, description } = pickedObject.id;
            console.log(`Hovered over: ${name} - ${description}`);
        }
    }, Cesium.ScreenSpaceEventType.MOUSE_MOVE);

    points.forEach(point => {
        const position = Cesium.Cartesian3.fromDegrees(point.coordinates.longitude, point.coordinates.latitude);

        const entity = viewer.entities.add({
            position: position,
            billboard: {
                image: 'https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/Red_dot.svg/120px-Red_dot.svg.png', // Marker image
                width: 16,
                height: 16
            },
            id: point, // Storing the entire point as the entity's id
            name: point.name,
            description: point.description // Adding description
        });
    });

    console.log('Markers added:', points);
}


// Function to fetch and process data from a given JSON file
function fetchData(file) {
    fetch(file)
      .then(response => response.json())
      .then(data => {
        console.log(`Fetched data from ${file}:`, data);
  
        // Example: Processing data based on the file name
        if (file === 'pointsOfInterest.json') {
          data.forEach(point => {
            console.log(`Name: ${point.name}, Latitude: ${point.coordinates.latitude}, Longitude: ${point.coordinates.longitude}`);
          });
        } else if (file === 'surfVessels.json') {
          data.forEach(vessel => {
            console.log(`Vessel ID: ${vessel.id}, Latitude: ${vessel.latitude}, Longitude: ${vessel.longitude}`);
          });
        } else if (file === 'subseaAssets.json') {
          data.forEach(asset => {
            console.log(`Asset ID: ${asset.id}, Latitude: ${asset.latitude}, Longitude: ${asset.longitude}`);
          });
        } else if (file === 'subseaPipelines.json') {
          data.forEach(pipeline => {
            console.log(`Pipeline ID: ${pipeline.id}, Start Latitude: ${pipeline.startLatitude}, End Latitude: ${pipeline.endLatitude}`);
          });
        }
  
        // You can use this data to add markers or visualizations in your Cesium app
      })
      .catch(error => {
        console.error(`Error fetching ${file}:`, error);
      });
  }
  
  // Fetch all the required JSON files
  fetchData('points_of_interest.json');
  fetchData('surfVessels.json');
  fetchData('subseaAssets.json');
  fetchData('subseaPipelines.json');
  

File: apiCaller.py
# import requests
# import json

# endpoints = {
#     'pointsOfInterest': 'https://elementz.rguhack.uk/pointsOfInterest',
#     'surfVessels': 'https://elementz.rguhack.uk/surfVessels',
#     'subseaAssets': 'https://elementz.rguhack.uk/subseaAssets',
#     'subseaPipelines': 'https://elementz.rguhack.uk/subseaPipelines'
# }

# def fetch_and_save_data(endpoint, filename):
#     try:
#         response = requests.get(endpoint)
#         # Check if the request was successful
#         if response.status_code == 200:
#             data = response.json()
#             # Write the data to a local JSON file
#             with open(filename, 'w') as outfile:
#                 json.dump(data, outfile, indent=4)
#             print(f"Data has been saved to '{filename}'")
#         else:
#             print(f"Failed to fetch data from {endpoint}. Status code: {response.status_code}")
#     except Exception as e:
#         print(f"An error occurred while fetching data from {endpoint}: {e}")

# # Fetch and save data from all endpoints
# for name, url in endpoints.items():
#     fetch_and_save_data(url, f'{name}.json')


import requests
import json

endpoints = {
    'pointsOfInterest': 'https://elementz.rguhack.uk/pointsOfInterest',
    'surfVessels': 'https://elementz.rguhack.uk/surfVessels',
    'subseaAssets': 'https://elementz.rguhack.uk/subseaAssets',
    'subseaPipelines': 'https://elementz.rguhack.uk/subseaPipelines'
}

def fetch_and_save_data(endpoint, filename, coordinates):
    try:
        response = requests.get(endpoint)
        # Check if the request was successful
        if response.status_code == 200:
            data = response.json()
            # Extract coordinates and add to the list
            for item in data:
                print(item)
                if 'coordinates' in item:
                    coords = item['coordinates']
                    if 'coordinates' in coords:
                        print("Subsea asset detected, recording coordinates.")
                        coordinates.append({
                            'latitude':coords.get('coordinates').get('latitude'),
                            'longitude':coords.get('coordinates').get('longitude')
                        })
                    elif coords is not None:
                        coordinates.append({
                            'latitude': coords.get('latitude'),
                            'longitude': coords.get('longitude')
                        })
                else:
                    if 'start_coordinates' in item:
                        print("Pipeline detected. Gathering both start and end point coordinates...")
                        coords = item['start_coordinates']
                        coordinates.append({
                            'latitude': coords.get('coordinates').get('latitude'),
                            'longitude': coords.get('coordinates').get('longitude')
                        })
                        coords = item['end_coordinates']
                        coordinates.append({
                            'latitude': coords.get('coordinates').get('latitude'),
                            'longitude': coords.get('coordinates').get('longitude')
                        })
                    # else {

                    # }
                    
            # Write the full data to a local JSON file
            with open(filename, 'w') as outfile:
                json.dump(data, outfile, indent=4)
            print(f"Data has been saved to '{filename}'")
        else:
            print(f"Failed to fetch data from {endpoint}. Status code: {response.status_code}")
    except Exception as e:
        print(f"An error occurred while fetching data from {endpoint}: {e}")

def save_coordinates_to_file(coordinates, filename):
    try:
        # Write the coordinates to a local JSON file
        with open(filename, 'w') as outfile:
            json.dump(coordinates, outfile, indent=4)
        print(f"Coordinates have been saved to '{filename}'")
    except Exception as e:
        print(f"An error occurred while saving coordinates: {e}")

# List to store all coordinates
coordinates = []

# Fetch and save data from all endpoints, and save coordinates separately
for name, url in endpoints.items():
    fetch_and_save_data(url, f'{name}.json', coordinates)

# After fetching data, save all coordinates to a new file
save_coordinates_to_file(coordinates, 'locations.json')

print(coordinates)
